### PR TITLE
Added rancher2_app resource and datasource. Go, test and doc files

### LINKED ISSUE #69: Add ability to provision applications within a project
Provisioning apps from Rancher or Helm charts would be super useful. In my case, I'm setting up our CI/CD environment and need to deploy the Jenkins helm chart within a project.

I can try to tackle this if it's a feature that's desired. If there's a workaround instead, I'll take that as well.

### PR DESCRIPTION
**Pull request by gabrielrcouto**
Fix #69

### CODE REVIEW

**Comment by rawmind0** [comment]
@gabrielrcouto , thanks for the PR.

Acceptance tests are not passing, `The argument "target_namespace" is required, but no definition was found`. `target_namespace` is not defined in the test files, could you please take a look??

**Comment by gabrielrcouto** [comment]
@rawmind0 fixed and refactored some details of the implementation 😄

File: rancher2/config.go [hunk 1]
```diff
@@ -1028,3 +1028,68 @@ func (c *Config) DeleteRegistry(registry interface{}) error {
...
+	if app == nil {
+		return fmt.Errorf("[ERROR] App can't be nil")
+	}
+
+	client, err := c.ProjectClient(app.ProjectID)
+	if err != nil {
+		return err
+	}
+	return client.App.Delete(app)
+}
```
Comment (rawmind0) [inline rancher2/config.go:1095]: `CreateApp`, `UpdateApp` and `DeleteApp` functions code should be included in `resourceRancher2AppCreate`, `resourceRancher2AppUpdate` and `resourceRancher2AppDelete` functions on `rancher2/resource_rancher2_app.go`. They also should include check for the state of the app. For example, when an App is created, it should be on `active` state.

File: rancher2/schema_app.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+		"project_id": &schema.Schema{
+			Type:        schema.TypeString,
+			Required:    true,
+			ForceNew:    true,
+			Description: "Project ID to add app",
+		},
+		"name": &schema.Schema{
+			Type:        schema.TypeString,
+			Required:    true,
+			ForceNew:    false,
```
Comment (rawmind0) [inline rancher2/schema_app.go:20]: `name` argument should be set as `ForceNew: true` due to app name couldn't be updated once is deployed.

File: rancher2/schema_app.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+		"annotations": &schema.Schema{
+			Type:        schema.TypeMap,
+			Optional:    true,
+			Computed:    true,
+			Description: "Annotations of the app",
+		},
+		"answers": &schema.Schema{
+			Type:        schema.TypeMap,
+			Optional:    true,
+			ForceNew:    false,
```
Comment (rawmind0) [inline rancher2/schema_app.go:44]: Default value for `ForceNew` is `false`, could you please remove its definition??

```
-			ForceNew:    false,
```

File: rancher2/schema_app.go [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+		"labels": &schema.Schema{
+			Type:        schema.TypeMap,
+			Optional:    true,
+			Computed:    true,
+			Description: "Labels of the app",
+		},
+		"values_yaml": &schema.Schema{
+			Type:        schema.TypeString,
+			Optional:    true,
+			ForceNew:    false,
```
Comment (rawmind0) [inline rancher2/schema_app.go:60]: Default value for `ForceNew` is `false`, could you please remove its definition??

```
-			ForceNew:    false,
```

File: rancher2/resource_rancher2_app.go [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+	log.Printf("[INFO] Updating App ID %s", id)
+
+	app, err := meta.(*Config).GetApp(id, projectID)
+	if err != nil {
+		return err
+	}
+
+	update := map[string]interface{}{
+		"name":            d.Get("name").(string),
+		"targetNamespace": d.Get("target_namespace").(string),
```
Comment (rawmind0) [inline rancher2/resource_rancher2_app.go:89]: As `name` and `target_namespace` arguments are set with `ForceNew: true`, they shouldn't be included in update map.

```
-		"name":            d.Get("name").(string),
-		"targetNamespace": d.Get("target_namespace").(string),
```

File: rancher2/resource_rancher2_app.go [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+		"externalId":      d.Get("external_id").(string),
```
Comment (rawmind0) [inline rancher2/resource_rancher2_app.go:90]: Update app `external_id` is an app upgrade. It should be done using `ActionUpgrade` function,  https://github.com/rancher/types/blob/master/client/project/v3/zz_generated_app.go#L91

File: rancher2/schema_app.go [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+			Description: "values.yaml file content of the app",
+		},
+	}
+
+	return s
+}
```
Comment (rawmind0) [inline rancher2/schema_app.go:66]: I'd add 2 additional fields, `revision_id` as computed and `force_upgrade` as optional bool to manage upgrade/rollback actions.

**Review by rawmind0 — CHANGES_REQUESTED** [review]

**Comment by rawmind0** [comment]
@gabrielrcouto , any update here??

**Comment by rawmind0** [comment]
@gabrielrcouto, done a new PR based in yours, thanks. New PR includes requested changes and some improvements on app update.

Superseded by PR #125

**Comment by gabrielrcouto** [comment]
Thank you very much @rawmind0 and sorry for the delay to answer, unfortunately I had no time in the last weeks to work on this PR.

**Pull request closed without merging**